init: pebble-sdk at 4.5

edanaher opened the pull request:
###### Motivation for this change

While Pebble has been bought by Fitbit and stopped selling new products, the watches are still around and available (and cheap!).  So this adds the pebble sdk to develop apps for it.

This is by far my most complex PR to date; there are definitely some rough edges and probably things that could be done better.  E.g., I used pypi2nix to generate the dependencies, since the sdk's requirements.txt wants specific versions which are often slightly older than the versions in nixpkgs (though I don't know if that's actually a requirement or just paranoia).  Also, the user-facing tooling is written in Python, but a C toolchain is also included, so we need to patch both binaries and python scripts;  I ended up doing that "manually" but would love to hear if there's a better way.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
init: pebble-sdk at 4.5

FRidh commented:
So these are the requirements:
```
libpebble2==0.0.26
enum34==1.0.4
httplib2==0.9.1
oauth2client==1.4.12
progressbar2==2.7.3
pyasn1==0.1.8
pyasn1-modules==0.0.6
pypng==0.0.17
pyqrcode==1.1
requests==2.7.0
rsa==3.1.4
pyserial==2.7
six==1.9.0
virtualenv==13.1.0
websocket-client==0.32.0
wheel==0.24.0
colorama==0.3.3
packaging==16.7
https://s3-us-west-2.amazonaws.com/pebble-sdk-homebrew/pypkjs-1.0.6.tar.gz
```
Apparently they have been fixed by something like `pip freeze`. For many of these dependencies there is absolutely no need to pin them. 

Please try and use the packages we have currently in `python-packages.nix`. Aside from 1 or 2 all of them are already packaged. I highly doubt the sdk won't work without exactly these versions.

FRidh commented:
Furthermore, unless the sdk includes importable modules (i.e., its a library itself) you might want to use `pythonEnv = python.withPackages(ps: dependencies)` here.

edanaher commented:
@FRidh Thanks for the input!  I'll try your advice.

Commit:
pebble-sdk: Add exec-phonesim.patch

Caught by travis-ci; nicely done!

Commit:
pebble-sdk: cleanups from code review

- Don't require exact versions of python packages
- Use existing python packages where appropriate
- Use pythonEnv to set up the python dependencies
- Other minor cleanups

Commit:
python-packages: Add libraries required for pebble-sdk

- peewee
- pypng
- progressbar2
- pyqrcode
- wsgiref

FRidh commented:
What is the status of this PR?

edanaher committed:
python-packages: Fix metadata for progressbar2

edanaher commented:
I cleaned it up a bit, and thought it was better; I just noticed that travis-ci failed because I used "bsd" for one of the licenses instead of "bsd3"; I've fixed that and re-pushed.

Aside from that, I think it's good, but would be happy to get more feedback if you have any suggestions.

edanaher commented:
I'm a bit confused by that failing travis-ci build; it's failing with the following 
```
error: a ‘aarch64-linux’ is required to build ‘/nix/store/yf3mids7kd7gwxbfpl2mkrd9wydl7h8j-binutils-2.28.drv’, but I am a ‘x86_64-linux’
```
But when I build locally (on x86_64), the sdk builds just fine, as does the nox-review command given in the PR checklist.  And if I try to run the command from travis-ci (travis-nox-review-pr.sh), it fails due to a missing /nixos/release.nix.

Can I get a hand either with possible causes of the error or with how to reproduce it?

7c6f434c commented:
Well, there is `nox` package with `nox-review wip --against upstream/master` where upstream is the remote name; and for evaluation (which currently seems a problem, too) there is `./maintainers/scripts/test-eval-release.sh`

c0bw3b commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,57 @@
+rec {
+  libpebble2 = python27Packages.buildPythonPackage {
+    name = "libpebble2-0.0.26";
+    src = pkgs.fetchurl {
```
You could use `fetchPypi` here

c0bw3b commented:
@edanaher do you still want this merged? If you could rebase it on master and resolve the conflicts, please. :)

And take a look at `fetchPypi` for libpebble2

edanaher commented:
Yeah... I never managed to figure out the travis failure, and it Works For Me for now, so I kind of gave up on the PR.  Maybe I'll take another look at it in the next few days.

c0bw3b commented:
The Travis failures were probably unrelated to your commits so don't worry about it.
BTW we just got rid of it. :) (replaced by a new bot/borg)

edanaher commented:
Whoops; didn't mean to push this yet.  It's rebased and conflicts are resolved, but still needs some cleanup from a few months of bitrot.

edanaher committed:
pebble-sdk: Cleanups from a few months of nixpkgs changes.

edanaher committed:
pebble-sdk: Use fetchPypi for libpebble2.

Thanks for the PR comment, c0bw3b.

edanaher committed:
pyqrcode: Fix license and use fetchPypi.

FRidh commented on pkgs/development/mobile/pebblesdk/default.nix:
```diff
@@ -0,0 +1,77 @@
+{ pkgs, stdenv, lib, fetchurl, makeWrapper
```
no `pkgs`

FRidh commented on pkgs/development/mobile/pebblesdk/default.nix:
```diff
@@ -0,0 +1,77 @@
+{ pkgs, stdenv, lib, fetchurl, makeWrapper
+, python27, python27Packages
```
either `python27` or `python27Packages`

FRidh commented on pkgs/development/mobile/pebblesdk/default.nix:
```diff
@@ -0,0 +1,77 @@
+    sha256 = "15yiypx9rnwyzsn4s4z1wmn1naw6mk7dpsiljmw3078ag66z1ca7";
+  };
+
+  pythonEnv = (python27.withPackages (ps: with ps; builtins.attrValues pebble-libs ++
```
this can go in a `let .. in`

FRidh commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,59 @@
+{ pkgs, python27Packages }:
```
No `pkgs`

FRidh commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,59 @@
+{ pkgs, python27Packages }:
+
+rec {
+  libpebble2 = python27Packages.buildPythonPackage rec {
```
I suggest using `with python27Packages;` at the top

FRidh commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,59 @@
+
+rec {
+  libpebble2 = python27Packages.buildPythonPackage rec {
+    name = "${pname}-${version}";
```
The `name` attribute is added by `buildPython*` and should therefore be removed.

FRidh commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,59 @@
+      sha256 = "16n69xxma7k8mhl8birdwa0fsqvf902g08s80mjb477s4dcxrvaz";
+    };
+
+    # Tries and fails to pull in pytest-mock
```
We have that test dependency, it can be added to `checkInputs`

FRidh commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,59 @@
+  };
+
+  pypkjs = python27Packages.buildPythonPackage {
+    name = "pypkjs-1.0.6";
```
The `name` attribute is added by `buildPython*` and should therefore be removed. Use `pname` and `version` instead.

FRidh commented on pkgs/development/mobile/pebblesdk/pebble-libs.nix:
```diff
@@ -0,0 +1,59 @@
+      wsgiref
+    ];
+
+    patchPhase = ''
```
`postPatch`

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6484,6 +6484,23 @@ in {
     };
   };
 
+  peewee = buildPythonPackage rec {
```
We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6484,6 +6484,23 @@ in {
   };
 
+  peewee = buildPythonPackage rec {
+    name = "peewee-2.8.8";
```
The `name` attribute is added by `buildPython*` and should therefore be removed.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6681,8 +6698,23 @@ in {
+
   pyphen = callPackage ../development/python-modules/pyphen {};
 
+  pypng = buildPythonPackage rec {
```
We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -16242,6 +16274,22 @@ in {
     };
   });
 
+  progressbar2 = buildPythonPackage (rec {
```
We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -18314,6 +18362,24 @@ in {
       };
     in if isPy3k then py3 else py2;
 
+  pyqrcode = buildPythonPackage rec {
```
We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -20519,6 +20585,20 @@ EOF
 
   magic-wormhole = callPackage ../development/python-modules/magic-wormhole { };
 
+  wsgiref = buildPythonPackage rec {
```
We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

FRidh commented:
Because of inactivity I will close this PR. When you do like to get this in, please make the required changes and open a new PR.
